h2/src/main/org/h2/command/Command.java

```diff
@@ -193,47 +194,50 @@ public ResultInterface executeQuery(int maxrows, boolean scrollable) {
                 // wait
             }
         }
-        synchronized (sync) {
-            session.setCurrentCommand(this, false);
-            try {
-                while (true) {
-                    database.checkPowerOff();
-                    try {
-                        ResultInterface result = query(maxrows);
-                        callStop = !result.isLazy();
-                        return result;
-                    } catch (DbException e) {
-                        start = filterConcurrentUpdate(e, start);
-                    } catch (OutOfMemoryError e) {
-                        callStop = false;
-                        // there is a serious problem:
-                        // the transaction may be applied partially
-                        // in this case we need to panic:
-                        // close the database
-                        database.shutdownImmediately();
-                        throw DbException.convert(e);
-                    } catch (Throwable e) {
-                        throw DbException.convert(e);
+        session.startStatementWithinTransaction();
+        session.setCurrentCommand(this, false);
```

Comment:
`setCurrentCommand()` should be inside `synchronized (sync)` because it resets fields in `Session` that are used inside execution of the command.